op-node/rollup/derive/channel.go

```diff
@@ -58,27 +53,48 @@ func (ch *Channel) AddFrame(frame Frame, l1InclusionBlock eth.L1BlockRef) error
 	if frame.ID != ch.id {
 		return fmt.Errorf("frame id does not match channel id. Expected %v, got %v", ch.id, frame.ID)
 	}
+	// These checks are specified and cannot be changed without a hard fork.
 	if frame.IsLast && ch.closed {
 		return fmt.Errorf("cannot add ending frame to a closed channel. id %v", ch.id)
 	}
 	if _, ok := ch.inputs[uint64(frame.FrameNumber)]; ok {
 		return DuplicateErr
 	}
-	// TODO: highest seen frame vs endFrameNumber
+	if ch.closed && frame.FrameNumber >= ch.endFrameNumber {
+		return fmt.Errorf("frame number (%d) is greater than or equal to end frame number (%d) of a closed channel", frame.FrameNumber, ch.endFrameNumber)
+	}
 
 	// Guaranteed to succeed. Now update internal state
 	if frame.IsLast {
 		ch.endFrameNumber = frame.FrameNumber
 		ch.closed = true
 	}
+	// Prune frames with a number higher than the closing frame number when we receive a closing frame
+	if frame.IsLast && ch.endFrameNumber < ch.highestFrameNumber {
+		// Do a linear scan over saved inputs instead of ranging over ID numbers
+		var idsToPrune []uint64
+		for id, prunedFrame := range ch.inputs {
+			if id >= uint64(ch.endFrameNumber) {
+				idsToPrune = append(idsToPrune, id)
+			}
+			ch.size -= uint64(len(prunedFrame)) + frameOverhead
+		}
+		for _, id := range idsToPrune {
+			delete(ch.inputs, id)
+		}
+		ch.highestFrameNumber = ch.endFrameNumber
+	}
+	// Update highest seen frame number after pruning
+	if frame.FrameNumber > ch.highestFrameNumber {
+		ch.highestFrameNumber = frame.FrameNumber
+	}
+
 	if ch.highestL1InclusionBlock.Number < l1InclusionBlock.Number {
 		ch.highestL1InclusionBlock = l1InclusionBlock
 	}
 	ch.inputs[uint64(frame.FrameNumber)] = frame.Data
 	ch.size += uint64(len(frame.Data)) + frameOverhead
```

Comment:
Perhaps a helper function like `Frame.Size()` would be useful and replace the inline casting + overhead

Reply:
added as a free function. I did have to change how some data was stored to keep a clean usage of it.